config/internal/warning.go

```diff
@@ -0,0 +1,64 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal // import "go.opentelemetry.io/collector/config/internal"
+
+import (
+	"net"
+	"strconv"
+	"strings"
+
+	"go.uber.org/zap"
+)
+
+func shouldWarn(endpoint string) bool {
+	switch {
+	case endpoint == ":":
+		// : (aka 0.0.0.0:0)
+		return true
+	case strings.HasPrefix(endpoint, ":"):
+		// :<port> (aka 0.0.0.0:<port>)
+		_, err := strconv.ParseInt(endpoint[1:], 10, 64)
+		if err != nil { // Probably invalid, don't warn
+			return false
+		}
```

Comment:
The switch in golang does not execute things in order. So you may hit any of these branches if `endpoint == ":"`. This does not look like very deterministic to me, so I would prefer you to fix this.

Reply:
Are you sure? [The spec](https://go.dev/ref/spec#Switch_statements) says
> the case expressions,[...] are evaluated left-to-right and top-to-bottom; the first one that equals the switch expression triggers execution of the statements of the associated case; the other cases are skipped. If no case matches and there is a "default" case, its statements are executed.

and then
> A missing switch expression is equivalent to the boolean value true. 

A quick check on the playground also seems to confirm this https://go.dev/play/p/KBynu4xGhKD

Maybe you were thinking about the `select` statement?

Reply:
Probably I did not have enough coffee :) You are correct..

I still believe that this code is easier to read:

```golang
func shouldWarn(endpoint string) bool {
	if endpoint == ":" {
		// : (aka 0.0.0.0:0)
		return true
	}

	if strings.HasPrefix(endpoint, ":"):
		// :<port> (aka 0.0.0.0:<port>)
		_, err := strconv.ParseInt(endpoint[1:], 10, 64)
		if err != nil { // Probably invalid, don't warn
			return false
		}
		return true
	}

	// <host>:<port>
	host, _, err := net.SplitHostPort(endpoint)
	if err != nil { // Probably invalid, don't warn.
		return false
	}

	ip := net.ParseIP(host)
	return ip != nil && ip.IsUnspecified()
}
```

Reply:
Done, PTAL